Apply dirmode to top-level cache directory, too

particleflux opened the pull request:
Due to the cache directory already created in the constructor, if it doesn't exist already it will be created with wrong permissions. The fluent setter for dirMode is only applied after the directory was created.

This re-applies the directory permissions once on `withDirectoryMode()` if they don't match the default.

Fixes #46

Also added a specific test case for checking the cacheDir permissions.

It would seem to me that doing this in the `withDirectoryMode()` setter is probably the best solution, since it won't require checking the mode on every set/get/gc, and only sets it once on init.



| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #46

vjik commented:
May be use another way: add second optional parameter to constructor `$directoryMode`?

particleflux commented:
> May be use another way: add second optional parameter to constructor `$directoryMode`?

That's definitely an option, though imho it would lead to confusing usage, as the fluent setter would not affect that top level dir,
and you have two ways to set the `directoryMode`. Maybe remove the setter, too?

> If value will be changed by withDirectoryMode, this will only affect newly created directories.

True, though that would always be the case, I doubt it makes sense anywhere (in code) to go over all existing subdirs and change the mode. The whole initialization flow with the `withDirectoryMode()` is only usable if it's called very early in application startup. Otherwise, on high traffic applications, accessing the cache might hit it in-between the constructor and the setter, and mix up modes completely.

vjik commented:
> Maybe remove the setter, too?

Mark as deprecated and remove in next major release.

particleflux commented:
Alright, removed the previous change to `withDirectoryMode()` and fixed the issue via constructor parameter instead.

I kept the new constant (even though it's technically no longer necessary, since the single use in constructor default param would be enough). The `directoryMode` property is now set before calling `createDirectoryIfNotExists()`, which fixes the issue, too.

Renamed the old test since it now tests a deprecated method, and added a new test case, which tests both subdirectory and top-level directory permissions.

vjik commented on src/FileCache.php:
```diff
@@ -50,6 +50,7 @@ final class FileCache implements CacheInterface
 {
     private const TTL_INFINITY = 31_536_000; // 1 year
     private const EXPIRATION_EXPIRED = -1;
+    private const DEFAULT_DIRMODE = 0775;
```
```suggestion
    private const DEFAULT_DIRECTORY_MODE = 0775;
```

vjik commented on src/FileCache.php:
```diff
@@ -74,7 +75,7 @@ final class FileCache implements CacheInterface
      * but read-only for other users.
      */
-    private int $directoryMode = 0775;
+    private int $directoryMode = self::DEFAULT_DIRMODE;
```
```suggestion
    private int $directoryMode = self::DEFAULT_DIRECTORY_MODE;
```

vjik commented on src/FileCache.php:
```diff
@@ -93,13 +94,18 @@ final class FileCache implements CacheInterface
      * @throws CacheException If failed to create cache directory.
      */
-    public function __construct(string $cachePath)
+    public function __construct(string $cachePath, int $directoryMode = self::DEFAULT_DIRMODE)
```
```suggestion
    public function __construct(string $cachePath, int $directoryMode = self::DEFAULT_DIRECTORY_MODE)
```

vjik commented on src/FileCache.php:
```diff
@@ -253,6 +259,8 @@ public function withFileMode(int $fileMode): self
      * This value will be used by PHP chmod() function. No umask will be applied.
      * Defaults to 0775, meaning the directory is read-writable by owner and group, but read-only for other users.
+     *
+     * @deprecated Use $directoryMode in the constructor instead
```
```suggestion
     * @deprecated Use `$directoryMode` in the constructor instead
```

vjik commented on tests/FileCacheTest.php:
```diff
@@ -443,6 +443,29 @@ public function testDirMode(): void
         $this->assertEquals('0777', $permissions);
     }
 
+    public function testDirMode(): void
```
```suggestion
    public function testDirectoryMode(): void
```

vjik commented on tests/FileCacheTest.php:
```diff
@@ -422,7 +422,7 @@ public function testFileMode(): void
     }
 
-    public function testDirMode(): void
+    public function testDirModeDeprecated(): void
```
```suggestion
    public function testDirectoryModeDeprecated(): void
```

vjik commented on CHANGELOG.md:
```diff
@@ -2,7 +2,7 @@
 ## 3.0.1 under development
 
-- no changes in this release.
+- Chg #69: Set directoryMode via constructor and deprecate `withDirectoryMode()`
```
```suggestion
- Chg #69: Add optional parameter `$directoryMode` to `FileCache` constructor and deprecate `withDirectoryMode()`
  method (@particleflux)
```

vjik commented:
@particleflux Thanks! Good job 👍
